File: frontend/android/ForkLift/app/src/main/java/sonic/ultra/forklift/messages/Messages.kt
package sonic.ultra.forklift.messages

data class ListConnectionsRequest(
        val type : String = "ListConnections"
)

data class ListConnectionsRepsonse(
        val type : String = "ListConnections",
        val locations: List<Location>
)

data class ListDevicesRequest(
        val type : String = "ListDevices",
        val start : String?,
        val end : String?,
        val location : Location?
)

data class ListDevicesResponse(
        val type : String = "ListDevices",
        val data : List<Device>
)

data class OverviewForDevicesRequest(
        val type: String = "ListOverview",
        val devices : List<Id>,
        val start : DateTimeString,
        val end : DateTimeString
)

data class OverviewForDevicesRepsonse(
        val type: String = "ListOverview",
        val data : List<OverviewEntry>
)

data class DataForDevicesRequest(
        val type: String = "ListData",
        val devices : List<Id>,
        val start : DateTimeString,
        val end : DateTimeString
)

data class DataForDevicesRepsonse(
        val type : String = "ListData",
        val data : List<DataEntry>
)



// AST

typealias Location = String
typealias Id = String
typealias DateTimeString = String?

data class Device(
        val id : Id,
        val name : String,
        val location : Location
)

data class OverviewEntry(
        val id : String,
        val name : String,
        val location : Location,
        val loadedTime : Long,
        val emptyTime : Long
)

data class DataEntry(
        val id : Id,
        val name : String,
        val location : Location,
        val loadedTimes : List<List<DateTimeString>>,
        val emptyTimes : List<List<DateTimeString>>
)

File: frontend/analytics/hanalytics/src/Messages.hs
{-# LANGUAGE DeriveGeneric, OverloadedStrings #-}

module Messages where


import GHC.Generics
import Data.Aeson
import Data.ByteString.Lazy

data DataEntry = DE String String String [[String]] [[String]] deriving (Generic, Show)

instance FromJSON DataEntry where
    parseJSON (Object v) = DE
        <$> v .: "id"
        <*> v .: "name"
        <*> v .: "location"
        <*> v .: "loadedTimes"
        <*> v .: "emptyTimes"
    parseJSON _ = undefined

data Device = D String String String deriving (Generic,Show)

instance FromJSON Device where
    parseJSON (Object v) = D
        <$> v .: "id"
        <*> v .: "name"
        <*> v .: "location"
    parseJSON _ = undefined

data OverviewEntry = 
    OE { id :: String
       , name :: String
       , location :: String
       , loadedTime :: Integer
       , emptyTime :: Integer
    } deriving (Generic, Show)

instance FromJSON OverviewEntry



data LLRequest = LLQ String deriving (Generic,Show)

data LLResponse = LLA String [String] deriving (Generic,Show)

instance ToJSON LLRequest where
    toEncoding (LLQ t) = pairs ("type" .= t)

instance FromJSON LLResponse where
    parseJSON (Object v) = do
        t  <- v .: "type"
        ls <- v .: "data"
        return (LLA t ls)
    parseJSON _ = undefined



data ORequest = OQ String [String] String String deriving (Generic,Show)

data OResponse = OA String [OverviewEntry] deriving (Generic,Show)

instance ToJSON ORequest where
    toEncoding (OQ t ds start end) = 
        pairs ("type" .= t <> "devices" .= ds <> "start" .= start <> "end" .= end)
instance FromJSON OResponse where
    parseJSON = withObject "OResponse" $ \v -> OA
        <$> v .: "type"
        <*> v .: "data"



data LDRequest = LDQ String String String String deriving (Generic,Show)

data LDResponse = LDA String [Device] deriving (Generic,Show)

instance ToJSON LDRequest where
    toEncoding (LDQ t start end l) = 
        pairs ("type" .= t <> "start" .= start <> "end" .= end <> "location" .= l)
instance FromJSON LDResponse where
    parseJSON = withObject "LDResponse" $ \v -> LDA
        <$> v .: "type"
        <*> v .: "data"


data DataRequest = DataQ String [String] String String deriving (Generic,Show)

data DataResponse = DataA String [DataEntry] deriving (Generic,Show)

instance ToJSON DataRequest where
    toEncoding (DataQ t ids start end) = 
        pairs ("type" .= t <> "devices" .= ids <> "start" .= start <> "end" .= end)
instance FromJSON DataResponse where
    parseJSON = withObject "DataEntry" $ \v -> DataA
        <$> v .: "type"
        <*> v .: "data"